driver/asana_driver.py

```diff
@@ -0,0 +1,185 @@
+        except asana.error.RateLimitEnforcedError, e:
+            raise RateLimitError( e.message )
+       
+        # configure client
+        asana.Client.DEFAULTS['page_size'] = 1000
+        return
+
+    def iter_tasks(self):
+        """
+        Returns a list of all tasks.
+        """
+        
+        for p in self.projects:
+            for t in self.client.tasks.find_all({ 'project': p['id'] }):
+                t['project'] = p
+                yield t
+                pass
+            pass
+        
+        return
+
+    def retrieve_data(self, doc):
+        """
+        Used to download content for a single doc. Typically this involves
+        passing the doc `external_id` attribute to the datasource SDK to
+        initialize a download. It is not mandatory that the only function
+        this performs is downloading. Gmail, for example, discovers
+        all the emails in a thread using the root email within `retrieve_data`.
+
+        :param doc: A dictionary describing a document.
+
+        :return `RetrieveDataResult` with the appropriate values populated.
+        """
+        external_id = doc['external_id']
+        subtype = doc.get('subtype')
+        
+        try:
+            if subtype == 'attachment':
+                attachment = self.client.attachments.find_by_id(attachment=external_id)
+                
+                url = attachment['download_url']
+                
+                r = requests.get(url)
+                
+                return RetrieveDataResult(data=r.content) # TODO: check if this okay for potentially large files
+            elif subtype is not None:
+                raise Exception( 'unexpected subtype: ' + repr(s) )
+            else:
+                task = self.client.tasks.find_by_id(task=external_id)
+                
+                return RetrieveDataResult(data=task['notes']) # Note: technically, a this could be empty
+        except asana.error.NotFoundError, e:
+            raise ServiceUnavailableError( e.message ) # TODO: doesn't seem like the best exception to raise
+        except asana.error.RateLimitEnforcedError, e:
+            raise RateLimitError( e.message )
+        return
+
+    def retrieve_metadata(self, milestone):
+        """
+        This function gathers the file manifest from the datasource. The only
+        focus here is the metadata, such as name, container, mimetype, etc. No
+        content should be downloaded here, that is reserved for `retrieve_data`.
+
+        :param milestone: A persistent dictionary unique to the butter
+            user/datasource combination. Example usage is to store a cursor
+            identifier that will allow the subsequent run to pick up indexing
+            files where it last left off.
+
+        :returns `RetrieveMetadataResult` with the appropriate values populated.
+        """
+        
+        if milestone is None: milestone = { 'step':'stage1' } # TODO: verify use of milestone
```

Comment:
With the case of a large asana project, this may hold cursor information. It looks like they use a hash value for offset. Incorporating this is necessary for larger projects,

https://asana.com/developers/documentation/getting-started/pagination